atom/src/catom.cpp

```diff
@@ -370,6 +371,105 @@ CAtom_sizeof( CAtom* self, PyObject* args )
     return PyLong_FromSsize_t( size );
 }
 
+PyObject*
+CAtom_getstate( CAtom* self )
+{
+    cppy::ptr stateptr = PyDict_New();
+    if ( !stateptr )
+        return PyErr_NoMemory();  // LCOV_EXCL_LINE
+
+    cppy::ptr selfptr(pyobject_cast(self), true);
+
+    // Copy __dict__ if present. Using hasattr / getattr makes it slower
+    // than the py version hence the _PyObject_GetDictPtr.
+    if ( PyObject** dict = _PyObject_GetDictPtr( selfptr.get() ) )
+    {
+        if ( PyDict_Update( stateptr.get(), dict[0] ) )
+            return 0;
+    }
+
+    // Copy __slots__ if present. This assumes copyreg._slotnames was called
+    // during AtomMeta's initialization
+    {
+        cppy::ptr typeptr = PyObject_Type(selfptr.get());
+        if (!typeptr)
+            return 0;
+        cppy::ptr slotnamesptr = typeptr.getattr("__slotnames__");
+        if (!slotnamesptr.get())
+            return 0;
+        if (!PyList_CheckExact(slotnamesptr.get()))
+            return cppy::system_error( "slot names" );
+        for ( Py_ssize_t i=0; i < PyList_GET_SIZE(slotnamesptr.get()); i++ )
+        {
+            PyObject *name = PyList_GET_ITEM(slotnamesptr.get(), i);
+            cppy::ptr value = selfptr.getattr(name);
+            if (!value || PyDict_SetItem(stateptr.get(), name, value.get()) )
+                return  0;
+        }
+    }
+
+    cppy::ptr membersptr = selfptr.getattr(atom_members);
+    if ( !membersptr || !PyDict_CheckExact( membersptr.get() ) )
+        return cppy::system_error( "atom members" );
+
+    PyObject *name, *member;
+    Py_ssize_t pos = 0;
+    while ( PyDict_Next(membersptr.get(), &pos, &name, &member) ) {
+        cppy::ptr should_gs = member_cast( member )->should_getstate( self );
+        if ( !should_gs ) {
+            return 0;
+        }
+        int test = PyObject_IsTrue( should_gs.get() );
+        if ( test == 1) {
+            PyObject *value = PyObject_GetAttr( selfptr.get(), name );
```

Comment:
Not sure if this is correct? Shouldn't it  just continue if test == 0 or return if  test== -1?

Reply:
Yes it was a bad rebase, should be fixed now.